nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/StandardOauth2AccessTokenProvider.java

```diff
@@ -0,0 +1,300 @@
+            } else {
+                try {
+                    refreshAccessDetails();
+                } catch (Exception e) {
+                    getLogger().info("Couldn't refresh access token.", e);
+                    acquireAccessDetails();
+                }
+            }
+        }
+
+        return accessDetails;
+    }
+
+    private void acquireAccessDetails() {
+        getLogger().debug("Getting a new access token.");
+
+        FormBody.Builder acquireTokenBuilder = new FormBody.Builder();
+
+        if (grantType.equals(RESOURCE_OWNER_PASSWORD_CREDENTIALS_GRANT_TYPE.getValue())) {
+            acquireTokenBuilder.add("grant_type", "password")
+                .add("username", username)
+                .add("password", password);
+        } else if (grantType.equals(CLIENT_CREDENTIALS_GRANT_TYPE.getValue())) {
+            acquireTokenBuilder.add("grant_type", "client_credentials");
+        }
+
+        if (clientId != null) {
+            acquireTokenBuilder.add("client_id", clientId);
+            acquireTokenBuilder.add("client_secret", clientSecret);
+        }
+
+        RequestBody acquireTokenRequestBody = acquireTokenBuilder.build();
+
+        Request acquireTokenRequest = new Request.Builder()
+            .url(authorizationServerUrl)
+            .post(acquireTokenRequestBody)
+            .build();
+
+        this.accessDetails = getAccessDetails(acquireTokenRequest);
+    }
+
+    private void refreshAccessDetails() {
+        getLogger().debug("Refreshing access token.");
+
+        FormBody.Builder refreshTokenBuilder = new FormBody.Builder()
+            .add("grant_type", "refresh_token")
+            .add("refresh_token", this.accessDetails.getRefreshToken());
+
+        if (clientId != null) {
+            refreshTokenBuilder.add("client_id", clientId);
+            refreshTokenBuilder.add("client_secret", clientSecret);
+        }
+
+        RequestBody refreshTokenRequestBody = refreshTokenBuilder.build();
+
+        Request refreshRequest = new Request.Builder()
+            .url(authorizationServerUrl)
+            .post(refreshTokenRequestBody)
+            .build();
+
+        this.accessDetails = getAccessDetails(refreshRequest);
+    }
+
+    private AccessToken getAccessDetails(Request newRequest) {
+        try {
+            Response response = httpClient.newCall(newRequest).execute();
+            String responseBody = response.body().string();
+            if (response.code() != 200) {
+                getLogger().error(String.format("Bad response from the server during oauth2 request:\n%s", responseBody));
+                throw new ProcessException(String.format("Got HTTP %d during oauth2 request.", response.code()));
+            }
+
+            AccessToken accessDetails = ACCESS_DETAILS_MAPPER.readValue(responseBody, AccessToken.class);
+
+            return accessDetails;
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
```

Comment:
Recommend including a message with this exception:
```suggestion
            throw new UncheckedIOException("OAuth2 Request Failed [Connection Problem]", e);
```